src/ray/core_worker/store_provider/plasma_store_provider.cc

```diff
@@ -16,72 +16,101 @@ Status CoreWorkerPlasmaStoreProvider::Put(const RayObject &object,
+  // objects are all fetched if timeout is -1.
   int num_attempts = 0;
   bool should_break = false;
   int64_t remaining_timeout = timeout_ms;
-  // Repeat until we get all objects.
+  int64_t batch_size = RayConfig::instance().worker_fetch_request_size();
   while (!unready.empty() && !should_break) {
     std::vector<ObjectID> unready_ids;
     for (const auto &entry : unready) {
+      if (unready_ids.size() == batch_size) {
+        break;
+      }
       unready_ids.push_back(entry.first);
     }
 
-    // For the initial fetch, we only fetch the objects, do not reconstruct them.
-    bool fetch_only = num_attempts == 0;
-    if (!fetch_only) {
-      // If fetch_only is false, this worker will be blocked.
-      was_blocked = true;
-    }
-
-    // TODO(zhijunfu): can call `fetchOrReconstruct` in batches as an optimization.
-    RAY_CHECK_OK(raylet_client_->FetchOrReconstruct(unready_ids, fetch_only, task_id));
+    RAY_CHECK_OK(raylet_client_->FetchOrReconstruct(unready_ids, /*fetch_only=*/false, task_id));
 
-    // Get the objects from the object store, and parse the result.
-    int64_t get_timeout;
+    int64_t timeout = std::max(
+		    RayConfig::instance().get_timeout_milliseconds(), int64_t(0.01 * unready.size()));
     if (remaining_timeout >= 0) {
-      get_timeout =
-          std::min(remaining_timeout, RayConfig::instance().get_timeout_milliseconds());
-      remaining_timeout -= get_timeout;
+      timeout = std::min(remaining_timeout, timeout);
+      remaining_timeout -= timeout;
       should_break = remaining_timeout <= 0;
-    } else {
-      get_timeout = RayConfig::instance().get_timeout_milliseconds();
     }
 
     std::vector<std::shared_ptr<RayObject>> result_objects;
-    RAY_RETURN_NOT_OK(
-        local_store_provider_.Get(unready_ids, get_timeout, task_id, &result_objects));
+    RAY_RETURN_NOT_OK(local_store_provider_.Get(unready_ids, timeout, task_id, &result_objects));
 
+    // Add successfully retrieved objects to the result list and remove them from unready.
+    uint64_t successes = 0;
     for (size_t i = 0; i < result_objects.size(); i++) {
       if (result_objects[i] != nullptr) {
+        successes++;
         const auto &object_id = unready_ids[i];
-        (*results)[unready[object_id]] = result_objects[i];
+        for (int idx : unready[object_id]) {
+          (*results)[idx] = result_objects[i];
+	}
         unready.erase(object_id);
         if (IsException(*result_objects[i])) {
           should_break = true;
         }
       }
     }
 
-    num_attempts += 1;
+    if (successes < unready_ids.size()) {
+      num_attempts += 1;
+    }
     WarnIfAttemptedTooManyTimes(num_attempts, unready);
   }
 
-  if (was_blocked) {
-    RAY_CHECK_OK(raylet_client_->NotifyUnblocked(task_id));
-  }
-
+  // Notify unblocked because we blocked when calling FetchOrReconstruct with fetch_only=false.
+  RAY_CHECK_OK(raylet_client_->NotifyUnblocked(task_id));
```

Comment:
should just `return raylet_client_->NotifyUnblocked(task_id);`, in order to propagate errors to Python/Java, instead of crashing.